flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/ForStGeneralMultiGetOperation.java

```diff
@@ -33,39 +37,169 @@
+            int readIoParallelism,
             Runnable subProcessFinished) {
         this.db = db;
         this.batchRequest = batchRequest;
         this.executor = executor;
         this.subProcessFinished = subProcessFinished;
+        this.readIoParallelism = readIoParallelism;
+        this.splitRequests = new ArrayList<>();
+        this.mapCheckRequests = new ArrayList<>();
+        classifyAndSplitRequests(splitRequests, mapCheckRequests);
     }
 
     @Override
     public CompletableFuture<Void> process() {
-        // TODO: Use MultiGet to optimize this implement
 
         CompletableFuture<Void> future = new CompletableFuture<>();
-
         AtomicReference<Exception> error = new AtomicReference<>();
         AtomicInteger counter = new AtomicInteger(batchRequest.size());
+
+        processOneByOne(mapCheckRequests, error, counter, future);
+        for (List<ForStDBGetRequest<?, ?, ?, ?>> getRequests : splitRequests) {
+            executor.execute(
+                    () -> {
+                        ReadOptions readOptions = new ReadOptions();
+                        readOptions.setReadaheadSize(0);
+                        List<byte[]> keys = new ArrayList<>(getRequests.size());
+                        List<ColumnFamilyHandle> columnFamilyHandles =
+                                new ArrayList<>(getRequests.size());
+
+                        for (int i = 0; i < getRequests.size(); i++) {
+                            ForStDBGetRequest<?, ?, ?, ?> request = getRequests.get(i);
+                            try {
+                                if (error.get() == null) {
+                                    byte[] key = request.buildSerializedKey();
+                                    keys.add(key);
+                                    columnFamilyHandles.add(request.getColumnFamilyHandle());
+                                } else {
+                                    completeExceptionallyRequest(
+                                            request,
+                                            "Error already occurred in other state request of the same group, failed the state request directly",
+                                            error.get());
+                                }
+                            } catch (IOException e) {
+                                error.set(e);
+                                completeExceptionallyRequest(
+                                        request,
+                                        "Error when execute ForStDb serialized get key",
+                                        e);
+                                future.completeExceptionally(e);
+                            }
+                        }
+                        if (error.get() != null) {
+                            if (subProcessFinished != null) {
+                                subProcessFinished.run();
+                            }
+                            return;
+                        }
```

Comment:
Not well readable. How about use `try...finally...` and put `subProcessFinished.run();` within the `finally` clause?